File: web-dashboard/hook-queue-worker.js
const fs = require("node:fs");
const path = require("node:path");

const PORT = Number(process.env.AGENT_SIGNAL_LIGHT_PORT || 8787);
const HOOK_URL = `http://127.0.0.1:${PORT}/hook?agent=cursor`;
const QUEUE_DIR = path.join(process.env.HOME || "", ".cursor", "hooks", "queue");
const ATTEMPTS_DIR = path.join(QUEUE_DIR, ".attempts");
const POLL_MS = 50;
const MAX_ATTEMPTS = 3;
const STALE_MS = 60_000;

function ensureDirs() {
  fs.mkdirSync(QUEUE_DIR, { recursive: true });
  fs.mkdirSync(ATTEMPTS_DIR, { recursive: true });
}

function attemptsFor(name) {
  const p = path.join(ATTEMPTS_DIR, `${name}.count`);
  try {
    return Number(fs.readFileSync(p, "utf8")) || 0;
  } catch {
    return 0;
  }
}

function bumpAttempts(name) {
  const p = path.join(ATTEMPTS_DIR, `${name}.count`);
  const next = attemptsFor(name) + 1;
  try {
    fs.writeFileSync(p, String(next), "utf8");
  } catch {}
  return next;
}

function clearAttempts(name) {
  const p = path.join(ATTEMPTS_DIR, `${name}.count`);
  try { fs.unlinkSync(p); } catch {}
}

async function postPayload(raw) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 1500);
  try {
    const response = await fetch(HOOK_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: raw,
      signal: controller.signal
    });
    if (!response.ok) {
      throw new Error(`hook post failed: ${response.status}`);
    }
  } finally {
    clearTimeout(timer);
  }
}

async function drainOnce() {
  let names;
  try {
    names = fs.readdirSync(QUEUE_DIR).filter((name) => name.endsWith(".json")).sort();
  } catch {
    return;
  }

  for (const name of names) {
    const filePath = path.join(QUEUE_DIR, name);
    let raw = "";
    try {
      const stat = fs.statSync(filePath);
      if (Date.now() - stat.mtimeMs > STALE_MS) {
        fs.unlinkSync(filePath);
        clearAttempts(name);
        continue;
      }
      raw = fs.readFileSync(filePath, "utf8");
      if (raw.trim()) {
        await postPayload(raw);
      }
      fs.unlinkSync(filePath);
      clearAttempts(name);
    } catch (error) {
      const tries = bumpAttempts(name);
      if (tries >= MAX_ATTEMPTS) {
        try { fs.unlinkSync(filePath); } catch {}
        clearAttempts(name);
        console.error(`[hook-queue] ${name}: dropping after ${tries} attempts (${error.message})`);
      }
    }
  }
}

async function main() {
  ensureDirs();
  console.log(`[hook-queue] draining ${QUEUE_DIR} -> ${HOOK_URL}`);
  for (;;) {
    try {
      await drainOnce();
    } catch (error) {
      console.error(`[hook-queue] ${error.message}`);
    }
    await new Promise((resolve) => setTimeout(resolve, POLL_MS));
  }
}

main();

File: web-dashboard/hook-queue-worker.py
#!/usr/bin/env python3
"""Hook queue worker (Linux side).

Reads JSON payloads written by the hook script from a queue directory,
POSTs them to the AgentCore-Light server via the SSH reverse tunnel,
and removes the files.

Uses only the Python standard library so it runs on machines without Node.
"""
from __future__ import annotations

import os
import sys
import time
import urllib.request
from pathlib import Path

HOOK_URL = os.environ.get(
    "AGENT_LIGHT_HOOK_URL",
    "http://127.0.0.1:18787/hook?agent=cursor",
)
QUEUE_DIR = Path(os.environ.get(
    "AGENT_LIGHT_QUEUE_DIR",
    os.path.expanduser("~/.cursor/hooks/queue"),
))
ATTEMPTS_DIR = QUEUE_DIR / ".attempts"
POLL_SECONDS = 0.05
STALE_SECONDS = 60.0
MAX_ATTEMPTS = 3


def _attempts_path(name: str) -> Path:
    return ATTEMPTS_DIR / f"{name}.count"


def attempts_for(name: str) -> int:
    try:
        return int(_attempts_path(name).read_text(encoding="utf-8")) or 0
    except Exception:
        return 0


def bump_attempts(name: str) -> int:
    nxt = attempts_for(name) + 1
    try:
        _attempts_path(name).write_text(str(nxt), encoding="utf-8")
    except Exception:
        pass
    return nxt


def clear_attempts(name: str) -> None:
    try:
        _attempts_path(name).unlink()
    except Exception:
        pass


def post_payload(raw: str) -> None:
    request = urllib.request.Request(
        HOOK_URL,
        data=raw.encode("utf-8"),
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    opener = urllib.request.build_opener(urllib.request.ProxyHandler({}))
    response = opener.open(request, timeout=2)
    response.read()
    if response.status >= 400:
        raise RuntimeError(f"hook post failed: {response.status}")


def drain_once() -> None:
    QUEUE_DIR.mkdir(parents=True, exist_ok=True)
    ATTEMPTS_DIR.mkdir(parents=True, exist_ok=True)
    for path in sorted(QUEUE_DIR.iterdir()):
        if path.suffix != ".json":
            continue
        name = path.name
        try:
            age = time.time() - path.stat().st_mtime
            if age > STALE_SECONDS:
                path.unlink()
                clear_attempts(name)
                continue
            raw = path.read_text(encoding="utf-8")
            if raw.strip():
                post_payload(raw)
            path.unlink()
            clear_attempts(name)
        except Exception as exc:  # noqa: BLE001
            tries = bump_attempts(name)
            if tries >= MAX_ATTEMPTS:
                try:
                    path.unlink()
                except Exception:
                    pass
                clear_attempts(name)
                print(
                    f"[hook-queue] {name}: dropping after {tries} attempts ({exc})",
                    file=sys.stderr,
                )


def main() -> None:
    QUEUE_DIR.mkdir(parents=True, exist_ok=True)
    print(f"[hook-queue] draining {QUEUE_DIR} -> {HOOK_URL}", file=sys.stderr)
    while True:
        try:
            drain_once()
        except Exception as exc:  # noqa: BLE001
            print(f"[hook-queue] {exc}", file=sys.stderr)
        time.sleep(POLL_SECONDS)


if __name__ == "__main__":
    main()
